src/awkward/types/uniontype.py

```diff
@@ -92,11 +93,19 @@ def __repr__(self):
         args = [repr(self._contents), *self._repr_args()]
         return "{}({})".format(type(self).__name__, ", ".join(args))
 
-    def __eq__(self, other):
-        if isinstance(other, UnionType):
-            return (
-                type_parameters_equal(self._parameters, other._parameters)
-                and self._contents == other._contents
+    def _is_equal_to(self, other, all_parameters: bool):
+        compare_parameters = (
+            parameters_are_equal if all_parameters else type_parameters_equal
+        )
+        return (
+            isinstance(other, UnionType)
+            and compare_parameters(self._parameters, other._parameters)
+            and len(self._contents) == len(other._contents)
+            and any(
+                all(
+                    this._is_equal_to(that, all_parameters)
+                    for this, that in zip(self._contents, contents)
+                )
+                for contents in permutations(other._contents)
             )
```

Comment:
This seems like the correct thing to do, but it's really not ideal to be invoking the permutations in order to do so. AFAICR, we have not suggested that the order of a union's contents matters, so we must check equality like this.

Reply:
Although records and unions have a canonical `contents` order that they preserve, it would be correct to say that two records or two unions are equal if they have the same set of `contents`, regardless of order (`zip(fields, contents)` in the case of records, which puts an ordering back in for tuples).

I was trying some examples with our new canonicalization rules to see if we'd have to distinguish between types like

```
A = union[option[X], Y]
```

and

```
B = union[X, option[Y]]
```

I found that an attempt to create

```
C = option[union[X, Y]]
```

forces you to go through `simplified` and makes

```
D = union[option[X], option[Y]]
```

but it's still possible to create an asymmetric case (`A` or `B`). How about if we squash that avenue as well? That is, an attempt to construct

```python
UnionArray(tags, index, [option_type_array, non_option_type_array])
```

would (with a future rule) fail and force you to use `UnionArray.simplified` (already required wherever the codebase makes `UnionArray`, and then `UnionArray.simplified` would wrap all the non-option type arrays in an `UnmaskedArray` if any one of the provided `contents` is option-type.

How about it? Then you'd have fewer cases to consider when checking equality of types: you wouldn't have to worry about whether `A` and `B` should be considered equivalent because they can't exist (they always be `D`).

This comment can become a new issue, if you like that idea.

Reply:
Hmm, this is an interesting thought. I had not thought about the fact that `?union[X, Y]`, `union[?X, Y]` and `union[X, ?Y]` are degenerate. We don't permit `?union`, so perhaps we _do_ want this rule, too. Especially as `UnmaskedArray` is very low overhead.

I'm in favour, and I'll create an issue from this comment.